File: 2024 March/Print matrix in diagonal pattern/main.py
# Link: https://www.geeksforgeeks.org/problems/print-matrix-in-diagonal-pattern/1


class Solution:
    def matrixDiagonally(self, mat):
        n = len(mat)
        d = {}
        for i in range(n):
            for j in range(n):
                if i + j not in d:
                    d[i + j] = []
                d[i + j].append(mat[i][j])
        ans = []
        for key in d:
            if key % 2 == 0:
                ans.extend(d[key][::-1])
            else:
                ans.extend(d[key])
        return ans


if __name__ == "__main__":
    obj = Solution()
    mat = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    print(obj.matrixDiagonally(mat))
    mat = [[1, 2], [3, 4]]
    print(obj.matrixDiagonally(mat))

File: 2024 March/Print matrix in diagonal pattern/main.cpp
// Link: https://www.geeksforgeeks.org/problems/print-matrix-in-diagonal-pattern/1

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<int> matrixDiagonally(vector<vector<int>> &mat)
    {
        map<int, vector<int>> m;
        for (int i = 0; i < mat.size(); i++)
        {
            for (int j = 0; j < mat.size(); j++)
            {
                m[i + j].push_back(mat[i][j]);
            }
        }

        vector<int> v;
        int k = 0;
        for (auto i : m)
        {
            if (k % 2 == 0 and k > 0)
            {
                reverse(i.second.begin(), i.second.end());
            }

            for (auto ele : i.second)
            {
                v.push_back(ele);
            }

            k++;
        }
        return v;
    }
};

int main()
{
    Solution obj;
    vector<vector<int>> mat = {{1, 2, 3},
                               {4, 5, 6},
                               {7, 8, 9}};
    vector<int> ans = obj.matrixDiagonally(mat);
    for (auto i : ans)
    {
        cout << i << " ";
    }
    cout << endl;

    mat = {{1, 2},
           {3, 4}};
    ans = obj.matrixDiagonally(mat);
    for (auto i : ans)
    {
        cout << i << " ";
    }
    cout << endl;

    return 0;
}